twister: ignore serial device SerialException.

chen-png opened the pull request:
twister: ignore serial device SerialException.

For some boards, like up_squared and ehl_crb, they need do a
power off/on operation and use bios to load and execute zephyr
test binary, **during the power off, it may meet this SerialException,
but the serial object still works well after power on, we can continue
to get all results from serial, so we don't need to close it,
it's ok to ignore this exception.**
BTW, even if we ignore this exception here, it doesn't influence twister
function, because if it was really caused by disconnection, the test would
still be judged as failed due to timeout and serial would be closed later
in main thread, it doesn't impact results.
On the contrary, close serial directly in the monitor serial thread, it
will cause later tests failed due to mismatch errors.

fixes #45009 

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

nashif commented on scripts/twister:
```diff
@@ -608,6 +608,9 @@ structure in the main Zephyr tree: boards/<arch>/<board_name>/""")
         """)
 
+    parser.add_argument("--monitor-serial-later", action="store_true",
+            help="Monitor device serial after running flash process.")
```
This does not explain well what is going on here, the option itself is also awkward. Can you please explain the situation when this is needed and why is it needed, i.e. why do we need a special option when it works without the option most of time?

chen-png replied:
btw, I am curious why we need to invoke ser.close() when met a **serial exception "device reports readiness to read but returned no data (device disconnected or multiple access on port?)**", it seems that this exception maybe caused by other reasons, not only serial disconnected.
```
 try:
            serial_line = ser.readline()
            except TypeError:
                pass
            except serial.SerialException:
                ser.close()
                break
```
because I found I could still got all expected results from serial even I did change as below:
```
            except serial.SerialException:
                pass
```

hakehuang replied:
in NXP platform, this will introduce another issue i.e. missing leading log.

chen-png commented:
this is mainly caused by SerialException, I opened an issue https://github.com/zephyrproject-rtos/zephyr/issues/45009 to give more details.

chen-png commented:
@hakehuang @PerMac @gopiotr Could you help take a look of this? thanks.

PerMac reviewed (dismissed):
Using the most generic `Exception` and then ignoring it with `pass` looks like an example from "mistakes to avoid" guide. I read you description of the issue, but couldn't fully follow on what is the root of it. Could you try to refine the except part to something more reasonable?

chen-png commented:
@PerMac I think no matter what exceptions we got from serial, we shouldn't close serial directly in the monitor serial thread, it will be closed  by main thread anyway finally,

>if ser.isOpen():
  ser.close()

because if we closed serial in monitor serial thread when met an exception, but now only got half results of test instance A, the main thread won't wait at `t.join(self.timeout)` and it will run next test instance B immediately, then when the serial is opened again, it will catch another half results of previous test instance A.

PerMac commented:
I think I start following. So the issue is that we start serial connection at higher level than monitoring serial thread. Therefore serial connection should be managed on this higher level, and shouldn't be closed inside the thread. Shouldn't then instead of ignoring the serial exception, have the exception passed to the higher level and make it be handled there?
I see more places in `monitor_serial` where the  connection can get close so I guess this also might be problematic. 

TBH I have some problems understanding the logic of the twister code responsible for serial connection. I am worried it might require some more thoughtful refactoring. E.g. I believe that serial connection should be open using context menager (`with` statement) which makes sure it is closed at `finally`. 

@chen-png Will you join the upcoming testing wg meeting? Maybe we could discuss the issue there? 
Anyway, I will dismiss my request for change. The code there is already very "hacky", and I could imagine, that your proposal might be the easiest solution to the existing problem. However, I am not convinced on giving +1, since I cannot understand the implication of your change.

Commit:
twister: ignore serial device SerialException.

For some boards, like up_squared and ehl_crb, they need do a
power off/on operation and use bios to load and execute zephyr
test binary, during the power off, it may meet this SerialException,
but the serial object still works well after power on, we can continue
to get all results from serial, so we don't need to close it,
it's ok to ignore this exception.
BTW, even if we ignore this exception here, it doesn't influence twister
function, because if it was really caused by disconnection, the test would
still be judged as failed due to timeout and serial would be closed later
in main thread, it doesn't impact results.
On the contrary, close serial directly in the monitor serial thread, it
will cause later tests failed due to mismatch errors.

Signed-off-by: Chen Peng1 <peng1.chen@intel.com>

nashif commented:
original issue for this code in twister: https://github.com/zephyrproject-rtos/zephyr/issues/13541
